pkg/dogstatsd/server.go

```diff
@@ -260,77 +251,107 @@ func nextMessage(packet *[]byte) (message []byte) {
 	return message
 }
 
-func (s *Server) parsePacket(packet *listeners.Packet, metricSamples []*metrics.MetricSample, events []*metrics.Event, serviceChecks []*metrics.ServiceCheck) ([]*metrics.MetricSample, []*metrics.Event, []*metrics.ServiceCheck) {
-	extraTags := s.extraTags
+func (s *Server) parsePackets(batcher *batcher, packets []*listeners.Packet) {
+	for _, packet := range packets {
+		originTags := findOriginTags(packet.Origin)
+		//log.Tracef("Dogstatsd receive: %s", packet.Contents)
+		for {
+			message := nextMessage(&packet.Contents)
+			if message == nil {
+				break
+			}
+			if s.Statistics != nil {
+				s.Statistics.StatEvent(1)
+			}
+			messageType := findMessageType(message)
 
-	log.Tracef("Dogstatsd receive: %s", packet.Contents)
-	if packet.Origin != listeners.NoOrigin {
-		originTags, err := tagger.Tag(packet.Origin, tagger.DogstatsdCardinality)
-		if err != nil {
-			log.Errorf(err.Error())
-		} else {
-			extraTags = append(extraTags, originTags...)
+			switch messageType {
+			case serviceCheckType:
+				serviceCheck, err := s.parseServiceCheckMessage(message)
+				if err != nil {
+					log.Errorf("Dogstatsd: error parsing service check: %s", err)
+					continue
+				}
+				serviceCheck.Tags = append(serviceCheck.Tags, originTags...)
+				batcher.appendServiceCheck(serviceCheck)
+			case eventType:
+				event, err := s.parseEventMessage(message)
+				if err != nil {
+					log.Errorf("Dogstatsd: error parsing event: %s", err)
+					continue
+				}
+				event.Tags = append(event.Tags, originTags...)
+				batcher.appendEvent(event)
+			case metricSampleType:
+				sample, err := s.parseMetricMessage(message)
+				if err != nil {
+					log.Errorf("Dogstatsd: error parsing metrics: %s", err)
+					continue
+				}
+				if s.debugMetricsStats {
+					s.storeMetricStats(sample.Name)
+				}
+				sample.Tags = append(sample.Tags, originTags...)
+				batcher.appendSample(sample)
+				if s.histToDist && sample.Mtype == metrics.HistogramType {
+					distSample := sample.Copy()
+					distSample.Name = s.histToDistPrefix + distSample.Name
+					distSample.Mtype = metrics.DistributionType
+					batcher.appendSample(*distSample)
+				}
+			}
 		}
 	}
+	batcher.flush()
+}
 
-	for {
-		message := nextMessage(&packet.Contents)
-		if message == nil {
-			break
-		}
+func (s *Server) parseMetricMessage(message []byte) (metrics.MetricSample, error) {
+	sample, err := parseMetricSample(message)
+	if err != nil {
+		dogstatsdMetricParseErrors.Add(1)
+		return metrics.MetricSample{}, err
+	}
+	metricSample := enrichMetricSample(sample, s.metricPrefix, s.metricPrefixBlacklist, s.defaultHostname)
+	metricSample.Tags = append(metricSample.Tags, s.extraTags...)
```

Comment:
Couldn't `enrichMetricSample` be part of the `*Server` type and take care of appending the `s.extraTags` to the metric? Same goes for Events and ServiceChecks.

You would actually have a simpler signature, something like:
```
metricSample := s.enrichMetricSample(sample)
```

Reply:
Make sense, I'll add that in the caching PR if that's ok